brainiak/funcalign/rsrm.py

```diff
@@ -0,0 +1,560 @@
+                    + gamma * np.sum(np.abs(S[i]))
+        return func
+
+    @staticmethod
+    def _update_outliers(X, W, R, subjs, gamma):
+        """Updates the outlier terms `S_i`.
+
+        Parameters
+        ----------
+
+        X : list of 2D arrays, element i has shape=[voxels_i, samples]
+            Each element in the list contains the fMRI data for alignment of
+            one subject.
+
+        W : list of array, element i has shape=[voxels_i, features]
+            The orthogonal transforms (mappings) :math:`W_i` for each subject.
+
+        R : array, shape=[features, samples]
+            The shared response.
+
+        subjs : int
+            The number of subjects.
+
+        gamma : float, default: 1.0
+            Regularization parameter for the sparseness of the individual
+            components.
+
+        Returns
+        -------
+
+        S : list of array, element i has shape=[voxels_i, samples]
+            The individual component :math:`S_i` for each subject.
+        """
+        S = []
+        for i in range(subjs):
+            S.append(RSRM._shrink(X[i] - W[i].dot(R), gamma))
+        return S
+
+    @staticmethod
+    def _init_outliers(subjs, voxels, TRs):
+        """Initializes the outlier terms `S_i` to empty (all zeros).
+
+        Parameters
+        ----------
+
+        subjs : int
+            The number of subjects.
+
+        voxels : list of int
+            A list with the number of voxels per subject.
+
+        TRs : int
+            The number of samples in the data.
+
+        Returns
+        -------
+
+        S : list of 2D array, element i has shape=[voxels_i, samples]
+            The individual component :math:`S_i` for each subject initialized
+            to zero.
+        """
+        return [np.zeros((voxels[i], TRs)) for i in range(subjs)]
+
+    @staticmethod
+    def _update_shared_response(X, S, W, subjs, features, TRs):
+        """Update the shared response `R`.
+
+        Parameters
+        ----------
+
+        X : list of 2D arrays, element i has shape=[voxels_i, samples]
+            Each element in the list contains the fMRI data for alignment of
+            one subject.
+
+        S : list of array, element i has shape=[voxels_i, samples]
+            The individual component :math:`S_i` for each subject.
+
+        W : list of array, element i has shape=[voxels_i, features]
+            The orthogonal transforms (mappings) :math:`W_i` for each subject.
+
+        subjs : int
+            The number of subjects.
+
+        features : int
+            The number of features in the model.
+
+        TRs : int
+            The number of samples in the data.
+
+        Returns
+        -------
+
+        R : array, shape=[features, samples]
+            The updated shared response.
+        """
+        R = np.zeros((features, TRs))
+        for i in range(subjs):
+            R += W[i].T.dot(X[i])
+            R -= W[i].T.dot(S[i])
```

Comment:
I feel like this is such a critical step in RSRM that it should be explained somewhere (that you are subtracting from the shared response a component fitted to each participant)

Reply:
Done